libafl/src/observers/cmp.rs

```diff
@@ -303,51 +278,45 @@ where
     fn cmp_map_mut(&mut self) -> &mut CM {
         self.cmp_map.as_mut()
     }
-
-    fn cmp_observer_data(&self) -> <M as CmpObserverMetadata<'a, CM>>::Data {
-        <M as CmpObserverMetadata<CM>>::Data::default()
-    }
 }
 
-impl<'a, CM, S, M> Observer<S> for StdCmpObserver<'a, CM, S, M>
+impl<'a, CM, I, S, M> Observer<I, S> for StdCmpObserver<'a, CM, M>
 where
-    CM: CmpMap + Serialize + DeserializeOwned,
-    S: UsesInput + Debug + HasMetadata,
+    CM: Serialize + DeserializeOwned + CmpMap,
+    S: HasMetadata,
     M: CmpObserverMetadata<'a, CM>,
 {
-    fn pre_exec(&mut self, _state: &mut S, _input: &S::Input) -> Result<(), Error> {
+    fn pre_exec(&mut self, _state: &mut S, _input: &I) -> Result<(), Error> {
         self.cmp_map.as_mut().reset()?;
         Ok(())
     }
 
-    fn post_exec(
-        &mut self,
-        state: &mut S,
-        _input: &S::Input,
-        _exit_kind: &ExitKind,
-    ) -> Result<(), Error> {
+    fn post_exec(&mut self, state: &mut S, _input: &I, _exit_kind: &ExitKind) -> Result<(), Error> {
         if self.add_meta {
-            self.add_cmpvalues_meta(state);
+
+            #[allow(clippy::option_if_let_else)] // we can't mutate state in a closure
+            let meta = state.metadata_or_insert_with(|| M::new_metadata());
+        
+            let map = self.cmp_map_mut();
```

Comment:
i really can't fix this one... @addisoncrump 

![image](https://github.com/user-attachments/assets/075fb2f1-baea-4278-abd9-8c6c1f9c8524)

Reply:
Imagine I have two traits, and both of them take the same type parameter. and if this type parameter is not in the struct that i'm implementing. 
then i cannot call the method implemented in the other trait ☹️ 
```
impl traitA<X> for MyStruct<P> {
 pub do_a() {
  // do_stuff
 }
}

impl traitB<X> for MyStruct<P> {
 pub do_b() {
  self.do_a(); // won't work
 }
}

```

Reply:
it actually works but it says
```
try using a fully qualified path to specify the expected types
```
so i have to specify its long name

Reply:
exactly this

https://stackoverflow.com/questions/78226308/type-annotations-needed-try-using-a-fully-qualified-path-to-specify-the-expecte

Reply:
Is this resolved? I think we discussed splitting the trait.

Reply:
resolved